rllib/agents/ppo/ppo_torch_policy.py

```diff
@@ -182,24 +209,27 @@ def update_kl(self, sampled_kl):
 
 class ValueNetworkMixin:
     def __init__(self, obs_space, action_space, config):
-        if config["use_gae"]:
-
-            def value(ob, prev_action, prev_reward, *state):
-                model_out, _ = self.model({
-                    SampleBatch.CUR_OBS: torch.Tensor([ob]).to(self.device),
-                    SampleBatch.PREV_ACTIONS: torch.Tensor([prev_action]).to(
-                        self.device),
-                    SampleBatch.PREV_REWARDS: torch.Tensor([prev_reward]).to(
-                        self.device),
-                    "is_training": False,
-                }, [torch.Tensor([s]).to(self.device) for s in state],
-                                          torch.Tensor([1]).to(self.device))
-                return self.model.value_function()[0]
-
-        else:
-
-            def value(ob, prev_action, prev_reward, *state):
-                return 0.0
+        with torch.no_grad():
```

Comment:
Does adding torch.no_grad() here do anything?  It's not applied when the defined value functions are run, although that may not be desired if there are times when you're calling value() when computing the vf_loss and need gradients to update the value network.  It would seem that the added torch.no_grad() to postprocess_ppo_gae() before value() is called would cover the cases when you don't want gradients.

Reply:
It's not a problem. Calling self._value is purely for fetching purposes. The loss function uses the `model.value_function()` op directly and this is differentiable.

Reply:
Got it, and probably why when originally putting torch.no_grad() inside value() learning wasn't adversely affected.  Looking at this one change in isolation I don't see any effect on the memory leak problem as torch.no_grad() is only in use during __init__(), not when value() is actually called as it's not inside value().  I'm not sure if it's your intention to keep torch.no_grad() outside of value() as doing so doesn't really change much as far as I can tell, but I guess it doesn't really hurt either with calls to value() currently still being protected by the torch.no_grad() added to postprocess_ppo_gae() before value() is called.

Reply:
Yes, you are absolutely right. The problem is not the `compute_advantages` method, but the `_value` call itself. I removed the no_grad around the `compute_advantages` and moved one into the `def value` and the leak goes away. A much nicer solution. And now it makes sense.
Either way, it's not intuitive that torch would produce all these gradient data that's never gc'd.

Reply:
This is much cleaner now with only one no_grad left. Thanks so much!
Btw, I'm curious what your use case for RL is? Would you mind sharing some information with us on how you guys use RLlib?
Are you on our ray Slack channel?

Reply:
@sven1977 We're looking for a deep RL framework for conducting research in RL that supports pytorch, and something where it's straight forward to have it work with our own custom environments and models as well as possibly policies/algorithms.